mountpoint-s3/tests/fuse_tests/read_test.rs

```diff
@@ -309,3 +309,44 @@ fn read_errors_test_s3() {
 fn read_errors_test_mock(prefix: &str) {
     read_errors_test(fuse::mock_session::new, prefix);
 }
+
+fn read_after_flush_test<F>(creator_fn: F)
+where
+    F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    const KEY: &str = "data.bin";
+    let (mount_point, _session, mut test_client) = creator_fn("read_after_flush_test", Default::default());
+
+    let mut rng = ChaChaRng::seed_from_u64(0x87654321);
+    let mut two_mib_body = vec![0; 2 * 1024 * 1024];
+    rng.fill_bytes(&mut two_mib_body);
+    test_client.put_object(KEY, &two_mib_body).unwrap();
+
+    let path = mount_point.path().join(KEY);
+    let mut f = open_for_read(path, true).unwrap();
+
+    let mut content = vec![0; 128];
+    f.read_exact(&mut content).unwrap();
+
+    let mut f_dup = f.try_clone().unwrap();
+
+    // Close the file. Triggers a flush on the file handle.
+    drop(f);
+
+    // Read using the duplicated instance (same underlying handle).
+    // Seek to the end of the file to avoid relying on the kernel cache.
```

Comment:
Out of curiosity, how do we know how much data will the kernel cache in this case? Do we know that it will cache only those 128 bytes? Should you use `O_DIRECT` and `O_SYNC` when opening the file?

Reply:
We're using big numbers far larger than we expect the Kernel to cache.

I think we'd avoid `O_DIRECT` since it also has semantic changes in MP itself (skipping caches, always checking with S3). `O_SYNC` is not something we've used, I think its rejected which is fine as I believe its used when writing only - and we cannot fulfil the guarantees for `O_SYNC`, since a write cannot be guaranteed to be persisted to S3 - only `fsync`.